jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/FilesTask.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.gradle;
+
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskAction;
+
+/**
+ * Print out changing source dependencies on a project.
+ *
+ * <p>Expected use: {@code ./gradlew _jibSkaffoldFiles -q} or {@code ./gradlew
+ * :<subproject>:_jibSkaffoldFiles -q}
+ */
+public class FilesTask extends DefaultTask {
+
+  /**
+   * Recursive function for printing out a project's artifacts. Calls itself when it encounters a
+   * project dependency.
+   *
+   * @param project the project to list the artifacts for
+   * @param projectDependencyJars the set of jar files associated with each project dependency
+   *     encountered
+   */
+  private static void listFilesForProject(Project project, Set<File> projectDependencyJars) {
```

Comment:
Returning a `Set` rather than passing a mutable `Set` as a parameter seemed more complicated initially, maybe I can look into it again.

Reply:
Would it be easier to read if we just pulled out all projectConfigurations in one step? Either recursive or some kind of iterative dfs/bfs. Then process each configuration after that?

Reply:
Hmm, do you mean pulling L109-120 out of the recursive function? Wouldn't that still involve maintaining a set to pull out the project dependencies?

Reply:
Yeah, you would, I mean the set is inevitable since you have to track it. And I think my point is different than Q's. So I probably shouldn't have continued on this stream.

But in the end, this function is doing two separate things (printing and collecting), and it might be a little easier to read if they were separate.